File: app/src/main/java/me/ykrank/s1next/viewmodel/FavouriteViewModel.java
package me.ykrank.s1next.viewmodel;

import android.databinding.ObservableField;
import android.support.v7.widget.PopupMenu;
import android.view.MenuItem;
import android.view.View;

import com.google.common.base.Supplier;

import io.reactivex.functions.Consumer;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.Favourite;
import me.ykrank.s1next.data.api.model.Thread;
import me.ykrank.s1next.data.event.FavoriteRemoveEvent;
import me.ykrank.s1next.view.activity.PostListActivity;
import me.ykrank.s1next.widget.EventBus;


public final class FavouriteViewModel {

    public final ObservableField<Favourite> favourite = new ObservableField<>();

    private final Supplier<Thread> threadSupplier = () -> {
        Thread thread = new Thread();
        Favourite favourite = this.favourite.get();
        thread.setId(favourite.getId());
        thread.setTitle(favourite.getTitle());
        return thread;
    };

    public Consumer<View> onBind() {
        return v -> PostListActivity.bindClickStartForView(v, threadSupplier);
    }

    public View.OnLongClickListener removeFromFavourites(final EventBus eventBus) {
        return v -> {
            PopupMenu popup = new PopupMenu(v.getContext(), v);
            popup.setOnMenuItemClickListener((MenuItem menuitem) -> {
                switch (menuitem.getItemId()) {
                    case R.id.menu_popup_remove_favourite:
                        eventBus.post(new FavoriteRemoveEvent(favourite.get().getFavId()));
                        return true;
                    default:
                        return false;
                }
            });
            popup.inflate(R.menu.popup_favorites);
            popup.show();
            return true;
        };
    }
}

File: app/src/main/java/me/ykrank/s1next/viewmodel/FavouriteViewModel.kt
package me.ykrank.s1next.viewmodel

import android.view.MenuItem
import android.view.View
import android.view.View.OnLongClickListener
import androidx.appcompat.widget.PopupMenu
import androidx.databinding.ObservableField
import androidx.lifecycle.LifecycleOwner
import com.github.ykrank.androidtools.widget.RxBus
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.Favourite
import me.ykrank.s1next.data.api.model.Thread
import me.ykrank.s1next.view.event.FavoriteRemoveEvent
import me.ykrank.s1next.view.page.post.postlist.PostListActivity.Companion.bindClickStartForView

class FavouriteViewModel(private val lifecycleOwner: LifecycleOwner) {
    val favourite = ObservableField<Favourite>()

    fun onBind(): Function1<View, Any> {
        return { v: View ->
            bindClickStartForView(
                v, lifecycleOwner
            ) {
                favourite.get()?.let {
                    val thread = Thread()
                    thread.id = it.id
                    thread.title = it.title
                    thread
                }
            }
        }
    }

    fun removeFromFavourites(rxBus: RxBus): OnLongClickListener {
        return OnLongClickListener { v: View ->
            val popup = PopupMenu(v.context, v)
            popup.setOnMenuItemClickListener { menuitem: MenuItem ->
                if (menuitem.itemId == R.id.menu_popup_remove_favourite) {
                    rxBus.post(FavoriteRemoveEvent(favourite.get()!!.favId))
                    return@setOnMenuItemClickListener true
                }
                false
            }
            popup.inflate(R.menu.popup_favorites)
            popup.show()
            true
        }
    }
}
